File: Modules/SEDO/EMS/SIE.Web.EMS.EquipRepair/EquipRepairs/Commands/SaveEquipRepairCommand.cs
using AngleSharp.Dom;
using SIE.Domain;
using SIE.Domain.Validation;
using SIE.EMS.Checks.Plans.ViewModels;
using SIE.EMS.EquipRepair.Controller;
using SIE.EMS.EquipRepair.EquipRepairs;
using SIE.EMS.EquipRepairs.Enums;
using SIE.Web.Command;
using System;

namespace SIE.Web.EMS.EquipRepair.EquipRepairs.Commands
{
    /// <summary>
    /// 保存报修
    /// </summary>
    [JsCommand("SIE.Web.EMS.EquipRepair.EquipRepairs.Commands.SaveEquipRepairCommand")]
    public class SaveEquipRepairCommand : ViewCommand
    {
        /// <summary>
        /// 执行(提示是否已存在对应设备的未完成的报修单)
        /// </summary>
        /// <param name="args"></param>
        /// <param name="scope"></param>
        /// <returns></returns>
        protected override object Excute(ViewArgs args, string scope)
        {
            var bill = args.Data.ToJsonObject<EquipRepairBill>();
            if (bill == null)
            {
                throw new ValidationException("该数据不是报修数据格式。".L10N());
            }
            if (bill.EquipAccountId == 0 && bill.SparePartId == 0)
            {
                throw new ValidationException("设备或备件不能空。".L10N());
            }
            if (bill.DeviceAbnormalId == 0 && string.IsNullOrEmpty(bill.DeviceAbnormalRemark))
            {
                throw new ValidationException("【故障现象】与【故障现象（备注）】至少有一个不为空。".L10N());
            }
            bill.RepairState = EquipRepairState.ApplyRepair;
            //因改动点检,保养,润滑,定检等页面的报修,故来源状态从对应页面传入,不在此固定.
            RT.Service.Resolve<RepairController>().GenerateRepair(bill);
            return true;
        }
    }
}

File: Modules/SEDO/EMS/SIE.Web.EMS.EquipRepair/EquipRepairs/Commands/SaveEquipRepairCommand.js
SIE.defineCommand('SIE.Web.EMS.EquipRepair.EquipRepairs.Commands.SaveEquipRepairCommand', {
    extend:'SIE.cmd.FormSave',
    meta: { text: "保存", group: "edit", iconCls: "icon-SaveEntity icon-blue" },
    canExecute: function (view) {
        return true;
    },
    execute: function (view, source) {
        var me = this;
        var data = view.getCurrent().data;
        var detailData = view.findChild("SIE.EMS.EquipRepair.EquipRepairs.EquipRepairBill").getCurrent().data;

        data.DeviceAbnormalId = detailData.DeviceAbnormalId;
        data.DeviceAbnormalRemark = detailData.DeviceAbnormalRemark;
        data.ProduceState = detailData.ProduceState;
        data.UrgentDegree = detailData.UrgentDegree;
        if (data.EquipAccountId == null && data.SparePartId == null) {
            SIE.Msg.showError("设备或备件不能空！".t());
            return false;
        }

        if (data.DeviceAbnormalId == null && data.DeviceAbnormalRemark.length <= 0) {
            SIE.Msg.showError("【故障现象】与【故障现象（备注）】至少有一个不为空！".t());
            return false;
        }
        if (data.RepairType == 0) {
            SIE.invokeDataQuery({
                type: "SIE.Web.EMS.EquipRepair.EquipRepairs.DataQuerys.EquipRepairDataQuery",
                method: "CheckPlanWithUnFinishRepairBill",
                params: [data.EquipAccountId],
                async: false,
                token: view.token,
                callback: function (res) {
                    if (res.Success) {
                        if (res.Result.length > 0) {
                            SIE.Msg.askQuestion(Ext.String.format("设备存在未完成的维修单，报修人{0},是否继续报修？".t(), res.Result), function () {
                                me.dosave(view, data);
                            });
                        }
                        else {
                            me.dosave(view, data);
                        }
                    } else {
                        SIE.Msg.showError(res.Message);
                    }
                }
            })
        }
        else {
            me.dosave(view, data);
        }
    },

    // 保存
    dosave: function (view, data) {
        var me = this;
        view.execute({
            data: data,
            callback: function (res) {
                if (res.Success) {
                    me.onSavedMsg(view);
                }
            }
        })
    },

    /**
    * override 重写保存后提示信息
    * @param {type} view
    * @param {type} res
    */
    onSavedMsg: function (view) {
        var isClose = false;
        view.getCurrent().markSaved();
        SIE.Msg.showInstantMessage('提交成功'.t(), '报修'.t(), 3, function () {
            isClose = true;
            CRT.Workbench.closeCurrentTab();
        });
        if (!isClose) {
            Ext.defer(function () {
                CRT.Workbench.closeCurrentTab();
            }, 3000);
        }
    }
});